flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/PythonKeyedCoProcessOperator.java

```diff
@@ -0,0 +1,314 @@
+                                        .getUserCodeClassLoader()
+                                        .asClassLoader()));
+    }
+
+    @Override
+    public void open() throws Exception {
+        InternalTimerService<VoidNamespace> internalTimerService =
+                getInternalTimerService("user-timers", VoidNamespaceSerializer.INSTANCE, this);
+        timerService = new SimpleTimerService(internalTimerService);
+        reusableInput = new Row(5);
+        reusableTimerData = new Row(5);
+
+        this.collector = new TimestampedCollector<>(output);
+        super.open();
+    }
+
+    @Override
+    public void processElement1(StreamRecord<Row> element) throws Exception {
+        bufferedTimestamp.offer(element.getTimestamp());
+        writeTimestampAndWatermark(reusableInput, element, timerService.currentWatermark());
+        writeInput1(reusableInput, reuseRow, element);
+
+        getRunnerInputTypeSerializer().serialize(reusableInput, baosWrapper);
+        pythonFunctionRunner.process(baos.toByteArray());
+        baos.reset();
+        elementCount++;
+        checkInvokeFinishBundleByCount();
+        emitResults();
+    }
+
+    @Override
+    public void processElement2(StreamRecord<Row> element) throws Exception {
+        bufferedTimestamp.offer(element.getTimestamp());
+        writeTimestampAndWatermark(reusableInput, element, timerService.currentWatermark());
+        writeInput2(reusableInput, reuseRow, element);
+
+        getRunnerInputTypeSerializer().serialize(reusableInput, baosWrapper);
+        pythonFunctionRunner.process(baos.toByteArray());
+        baos.reset();
+        elementCount++;
+        checkInvokeFinishBundleByCount();
+        emitResults();
+    }
+
+    @Override
+    public void emitResult(Tuple2<byte[], Integer> resultTuple) throws Exception {
+        byte[] rawResult = resultTuple.f0;
+        int length = resultTuple.f1;
+        if (PythonOperatorUtils.endOfLastFlatMap(length, rawResult)) {
+            bufferedTimestamp.poll();
+        } else {
+            bais.setBuffer(rawResult, 0, length);
+            Row runnerOutput = getRunnerOutputTypeSerializer().deserialize(baisWrapper);
+            if (runnerOutput.getField(0) != null) {
+                registerTimer(runnerOutput);
+            } else {
+                collector.setAbsoluteTimestamp(bufferedTimestamp.peek());
+                collector.collect((OUT) runnerOutput.getField(3));
+            }
+        }
+    }
+
+    @Override
+    public TypeInformation<OUT> getProducedType() {
+        return outputTypeInfo;
+    }
+
+    @Override
+    public void onEventTime(InternalTimer<Row, VoidNamespace> timer) throws Exception {
+        bufferedTimestamp.offer(timer.getTimestamp());
+        processTimer(false, timer);
+    }
+
+    @Override
+    public void onProcessingTime(InternalTimer<Row, VoidNamespace> timer) throws Exception {
+        bufferedTimestamp.offer(Long.MIN_VALUE);
```

Comment:
I think using `null` is a bit more reasonable than Long.MIN_VALUE, and we need to change the logic in `emitResult` accordingly.